toolchain/check/function.cpp

```diff
@@ -0,0 +1,214 @@
+    prev_param_ref =
+        context.insts().Get(prev_param_ref.As<SemIR::AnyBindName>().value_id);
+  }
+
+  CARBON_CHECK(new_param_ref.Is<SemIR::Param>()) << new_param_ref;
+  CARBON_CHECK(prev_param_ref.Is<SemIR::Param>()) << prev_param_ref;
+
+  auto new_param = new_param_ref.As<SemIR::Param>();
+  auto prev_param = prev_param_ref.As<SemIR::Param>();
+  return new_param.name_id == prev_param.name_id;
+}
+
+// Returns true if two param refs are equal.
+static auto AreParamsEqual(Context& context, SemIR::InstId new_decl_id,
+                           SemIR::InstBlockId new_param_refs_id,
+                           SemIR::InstId prev_decl_id,
+                           SemIR::InstBlockId prev_param_refs_id,
+                           llvm::StringLiteral param_diag_label) -> bool {
+  // This will often occur for empty params.
+  if (new_param_refs_id == prev_param_refs_id) {
+    return true;
+  }
+  const auto new_param_ref_ids = context.inst_blocks().Get(new_param_refs_id);
+  const auto prev_param_ref_ids = context.inst_blocks().Get(prev_param_refs_id);
+  if (new_param_ref_ids.size() != prev_param_ref_ids.size()) {
+    CARBON_DIAGNOSTIC(FunctionSignatureParamCountMismatch, Error,
+                      "Function declared with {0}{1} parameter(s).", int32_t,
+                      llvm::StringLiteral);
+    CARBON_DIAGNOSTIC(FunctionSignatureParamCountPrevious, Note,
+                      "Function previously declared with {0}{1} parameter(s).",
+                      int32_t, llvm::StringLiteral);
+    context.emitter()
+        .Build(new_decl_id, FunctionSignatureParamCountMismatch,
+               new_param_ref_ids.size(), param_diag_label)
+        .Note(prev_decl_id, FunctionSignatureParamCountPrevious,
+              prev_param_ref_ids.size(), param_diag_label)
+        .Emit();
+    return false;
+  }
+  for (auto [index, new_param_ref_id, prev_param_ref_id] :
+       llvm::enumerate(new_param_ref_ids, prev_param_ref_ids)) {
+    if (!IsParamEqual(context, new_param_ref_id, prev_param_ref_id)) {
+      CARBON_DIAGNOSTIC(FunctionSignatureParamMismatch, Error,
+                        "Declaration of{1} parameter {0} doesn't match.",
+                        int32_t, llvm::StringLiteral);
+      CARBON_DIAGNOSTIC(FunctionSignatureParamPrevious, Note,
+                        "Previous declaration of{1} parameter {0}.", int32_t,
+                        llvm::StringLiteral);
+      context.emitter()
+          .Build(new_param_ref_id, FunctionSignatureParamMismatch,
+                 new_param_ref_ids.size(), param_diag_label)
+          .Note(prev_param_ref_id, FunctionSignatureParamPrevious,
+                prev_param_ref_ids.size(), param_diag_label)
+          .Emit();
+      return false;
+    }
+  }
+  return true;
+}
+
+// Returns true if the provided function signatures are equal, in the sense that
+// declarations can be merged.
+static auto IsFunctionSignatureEqual(Context& context,
```

Comment:
I think it'd be useful to use the "match" / "agree" terminology from the design here (https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/generics/details.md#matching-and-agreeing). Generally, when we say two declarations "match", we mean they're similar enough that they must either declare the same thing or be an error, and when we say they "agree" we mean that they're similar enough that they validly declare the same thing. So I think what we're checking here is whether two function signatures agree. (For now we're saying function declarations match if they have the same name; maybe overloading will change that, or maybe not.)

Reply:
Done, although noting that documentation, it might've been helpful to provide an explanation of what it means to agree but not match. It *sounds* like the only difference is that agreeing is simply "names are optional but if present must match"  versus matching being "names must match", but maybe I'm missing something?

Reply:
I think we're requiring matching, since names must match, and param names must match.

Reply:
Also, "MatchParam" because I can't think of a good name format that would allow for "Match" and "Agree" to be interchangeable. "MatchParam" is maybe "ParamAgrees"; "IsParamMatch" would be "DoesParamAgree".

Reply:
See also #1132 and #2477. We don't have a design update for functions yet, but the current direction is:

- Two function declarations match if they have the same name.
- Two function declarations agree if they have the same declaration syntax with names resolving to the same thing, except that parameter names can be omitted, parentheses can be added / removed if they don't change anything, and aliases can be expanded in one but not both declarations.

And the general rule is that if two declarations match, then it's an error if they don't agree.

(I don't think we need to get all the corner case details right in this PR; it's probably better to iterate there once this is in place.)

But the thing we're checking here is agreement, not matching.